### ISSUE TITLE
Rule idea: detect code duplication

### ISSUE THREAD

**Issue by mik01aj** [eslint/eslint#3957]
I'd like eslint to produce a warning when I have pieces of duplicated code.

Precisely, I'd like no *n*- (default 3) line chunk to be repeated more then *k* (default 2) times. If it happens, the *k+1*-th occurrence should trigger a warning: "These *n* lines are duplicated for *k+1*-th time".

Quite possibly we will also need to only take into account lines containing alpha-numeric characters, because we don't want to detect chunks like:

```
    return this;
},
<empty line>
```

This rule does not prevent errors, but helps maintain overall code quality. I think that (in its simplest version) it's generic enough to be useful for any code, regardless of JS dialect or framework.

Do you think it's a good idea to have this rule in core ESLint, or shall it be a plugin?

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by lo1tuma** [comment]
ESLint rules only operate on the context of a single file. So you won’t be able to detect duplicated code across multiple files.

I suggest to use tools like [`jscpd`](https://github.com/kucherenko/jscpd) or [`jsinspect`](https://github.com/danielstjules/jsinspect).

**Comment by mik01aj** [comment]
Wow, both `jscpd` and `jsinspect` seem awesome! I didn't know them. I will most definitely try them both.

But in the meanwhile, having a simple check like this, in scope of a single file, would still make sense to me.

**Comment by platinumazure** [comment]
@mik01aj Even if the project team doesn't think your rule is a good idea, there is nothing stopping you from writing a plugin which includes this as a rule. How about giving that a shot? Let us know if you need help getting started.

**Comment by nzakas** [comment]
As @lo1tuma suggested, this isn't an area where ESLint is the best tool, as each file is looked at separately.
